src/libraries/System.IO.Compression/src/System/IO/Compression/WinZipAesStream.cs

```diff
@@ -0,0 +1,744 @@
+
+        public WinZipAesStream(Stream baseStream, ReadOnlyMemory<char> password, bool encrypting, int keySizeBits = 256, long totalStreamSize = -1, bool leaveOpen = false)
+        {
+            ArgumentNullException.ThrowIfNull(baseStream);
+
+            _baseStream = baseStream;
+            _encrypting = encrypting;
+            _keySizeBits = keySizeBits;
+            _totalStreamSize = totalStreamSize; // Store the total size
+            _leaveOpen = leaveOpen;
+#pragma warning disable CA1416 // HMACSHA1 is available on all platforms
+            _aes = Aes.Create();
+#pragma warning restore CA1416
+            _aes.Mode = CipherMode.ECB;
+            _aes.Padding = PaddingMode.None;
+
+#pragma warning disable CA1416 // HMACSHA1 available on all platforms ?
+#pragma warning disable CA5350 // Do Not Use Weak Cryptographic Algorithms ?
+            _hmac = new HMACSHA1();
+#pragma warning restore CA5350 // Do Not Use Weak Cryptographic Algorithms
+#pragma warning restore CA1416
+
+            Array.Clear(_counterBlock, 0, 16);
+            _counterBlock[0] = 1;
+
+            if (_totalStreamSize > 0)
+            {
+                int saltSize = _keySizeBits / 16;
+                int headerSize = saltSize + 2; // Salt + Password Verifier
+                const int hmacSize = 10; // 10-byte HMAC
+
+                _encryptedDataSize = _totalStreamSize - headerSize - hmacSize;
+                _encryptedDataRemaining = _encryptedDataSize;
+            }
+            else
+            {
+                _encryptedDataSize = -1;
+                _encryptedDataRemaining = -1;
+            }
+
+            if (_encrypting)
+            {
+                // 8 for AES-128, 12 for AES-192, 16 for AES-256
+                int saltSize = _keySizeBits / 16;
+                _salt = new byte[saltSize];
+                RandomNumberGenerator.Fill(_salt);
+
+                DeriveKeysFromPassword(password, _salt);
+
+                Debug.Assert(_hmacKey is not null, "HMAC key should be derived");
+                _hmac.Key = _hmacKey!;
+                InitCipher();
+            }
+            else
+            {
+                ReadHeader(password);
+            }
+        }
+
+        private void DeriveKeysFromPassword(ReadOnlyMemory<char> password, byte[] salt)
+        {
+            byte[] passwordBytes = Encoding.UTF8.GetBytes(password.ToArray());
+
+            try
+            {
+                // AES key size + HMAC key size (same as AES key) + password verifier (2 bytes)
+                int keySizeInBytes = _keySizeBits / 8;
+                int totalKeySize = keySizeInBytes + keySizeInBytes + 2;
+
+                // WinZip AES uses SHA1 for PBKDF2 with 1000 iterations per spec
+                byte[] derivedKey = Rfc2898DeriveBytes.Pbkdf2(
+                passwordBytes,
+                salt,
+                1000,
+                HashAlgorithmName.SHA1,
+                totalKeySize);
```

Comment:
You can avoid allocating temporary array here by using stackalloc and using the Pbkdf2 overload that outputs the data into a Span<byte>